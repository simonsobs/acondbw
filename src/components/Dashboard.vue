<template>
  <div class="dashboard" style="position: relative;">
    <v-data-table
      v-if="items"
      :headers="headers"
      :items="items"
      :loading="state == State.LOADING"
      disable-sort
      hide-default-footer
      @click:row="clickRow"
      class="elevation-1"
    >
      <template v-slot:item.plural="{ item }">
        <router-link :to="{ name: 'ProductList', params: { productTypeName: item.name } }">
          <span class="capitalize font-weight-bold primary--text">{{ item.plural }}</span>
        </router-link>
      </template>
    </v-data-table>
    <v-alert v-if="error" type="error" style="width: 100%;">{{ error }}</v-alert>
    <dev-tool-loading-state-overriding-menu @state="devtoolState = $event"></dev-tool-loading-state-overriding-menu>
  </div>
</template>

<script>
import ALL_PRODUCTS_TYPES from "@/graphql/AllProductTypes.gql";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/DevToolLoadingStateOverridingMenu";

export default {
  name: "Dashboard",
  components: {
    DevToolLoadingStateOverridingMenu
  },
  data: () => ({
    edges: null,
    headers: [
      { text: "Product type", value: "plural" },
      { text: "Number of products", align: "end", value: "products.totalCount" }
    ],
    error: null,
    devtoolState: null,
    State: State
  }),
  apollo: {
    edges: {
      query: ALL_PRODUCTS_TYPES,
      update: data =>
        data.allProductTypes ? data.allProductTypes.edges : null,
      result(result) {
        this.error = result.error ? result.error : null;
      }
    }
  },
  computed: {
    state() {
      if (this.devtoolState) {
        return this.devtoolState;
      }

      if (this.loading) {
        return State.LOADING;
      } else if (this.error) {
        return State.ERROR;
      } else if (this.edges) {
        if (this.edges.length) {
          return State.LOADED;
        } else {
          return State.EMPTY;
        }
      } else {
        return State.NONE;
      }
    },
    loading() {
      return this.$apollo.queries.edges.loading;
    },
    items() {
      if (this.state == State.NONE) {
        return null;
      } else if (this.state == State.EMPTY) {
        return [];
      } else {
        return this.edges ? this.edges.map(edge => edge.node) : null;
      }
    }
  },
  methods: {
    clickRow(item) {
      this.$router.push({
        name: "ProductList",
        params: { productTypeName: item.name }
      });
    }
  },
  watch: {
    devtoolState: function() {
      this.error =
        this.devtoolState == State.ERROR ? "Error from Dev Tools" : null;
    }
  }
};
</script>

<style scoped>
.capitalize {
  text-transform: capitalize;
}

.dashboard >>> tbody tr :hover {
  cursor: pointer;
}
</style>
