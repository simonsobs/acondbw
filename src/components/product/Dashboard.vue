<template>
  <div class="dashboard px-5">
    <v-data-table
      :headers="headers"
      :items="items"
      :loading="loading"
      :items-per-page="-1"
      @click:row="clickRow"
    >
      <template v-slot:top>
        <v-alert v-if="error" variant="tonal" type="error" :text="error">
        </v-alert>
      </template>
      <template v-slot:[`item.plural`]="{ item }">
        <router-link
          :to="{
            name: 'ProductList',
            params: { productTypeName: item.raw.name },
          }"
          class="capitalize font-weight-bold text-primary text-decoration-none"
          v-text="item.raw.plural"
        >
        </router-link>
      </template>
      <template #bottom></template>
    </v-data-table>
    <dev-tool-loading-state-menu
      top="-30px"
      right="10px"
      v-model="devtoolState"
    >
    </dev-tool-loading-state-menu>
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from "vue";
import { useRouter } from "vue-router";

import { useAllProductTypesQuery } from "@/generated/graphql";

import { useQueryState } from "@/utils/query-state";

const router = useRouter();

const query = useAllProductTypesQuery();
type Query = typeof query;

function readEdges(query: Query) {
  const edgesAndNulls = query.data?.value?.allProductTypes?.edges;
  if (!edgesAndNulls) return [];
  return edgesAndNulls.flatMap((e) => (e ? [e] : []));
}

function readNodes(query: Query) {
  return readEdges(query).flatMap((e) => (e.node ? e.node : []));
}

function isEmpty(query: Query) {
  return readNodes(query).length === 0;
}

const headers = ref([
  { title: "Product type", key: "plural" },
  {
    title: "Number of products",
    align: "end",
    key: "products.totalCount",
  },
]);

const queryState = useQueryState(query, { isEmpty });
const { loading, loaded, empty, error, devtoolState } = queryState;

const items = computed(() => (empty.value ? [] : readNodes(query)));

function clickRow(event: Event, { item }) {
  router.push({
    name: "ProductList",
    params: { productTypeName: item.selectable.name },
  });
}
</script>

<style scoped>
.capitalize {
  text-transform: capitalize;
}

.dashboard {
  position: relative;
}

/* :deep() selects child components in scoped styles */
.dashboard:deep(tbody tr :hover) {
  cursor: pointer;
}
</style>
