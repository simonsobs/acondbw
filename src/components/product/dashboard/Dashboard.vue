<template>
  <div class="dashboard px-5">
    <v-data-table
      :headers="headers"
      :items="items"
      :loading="loading"
      :items-per-page="-1"
      @click:row="clickRow"
    >
      <template v-slot:top>
        <v-alert v-if="error" variant="tonal" type="error">
          {{ error }}
        </v-alert>
        <refresh-button :disabled="loading" @refresh="refresh">
        </refresh-button>
      </template>
      <template v-slot:[`item.plural`]="{ item }">
        <router-link
          :to="item.raw.to"
          class="text-capitalize font-weight-bold text-primary text-decoration-none"
          v-text="item.raw.plural"
        >
        </router-link>
      </template>
      <template #bottom></template>
    </v-data-table>
    <dev-tool-checkboxes top="-30px" right="10px" v-model="override">
    </dev-tool-checkboxes>
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from "vue";
import { useRouter } from "vue-router";
import { useAllProductTypesQuery } from "@/graphql/codegen/generated";

import RefreshButton from "./RefreshButton.vue";

import { useQueryResponse, useConnection } from "./query";
import { useOverride } from "./override";
import { useRefreshOnMutation } from "./refresh";

const queryResponse = useAllProductTypesQuery();
const connection = computed(() => queryResponse.data?.value?.allProductTypes);

const query = {
  ...useQueryResponse(queryResponse),
  ...useConnection(connection),
};

const { override, loading, error, nodes } = useOverride(query);

const refresh = query.refresh;
useRefreshOnMutation(refresh);

const headers = ref([
  { title: "Product type", key: "plural" },
  {
    title: "Number of products",
    align: "end" as const,
    key: "nProducts",
  },
]);

const items = computed(() =>
  nodes.value.map((node) => ({
    name: node.name,
    plural: node.plural,
    nProducts: node.products?.totalCount || 0,
    to: {
      name: "ProductList",
      params: { productTypeName: node.name },
    },
  }))
);

const router = useRouter();

function clickRow(event: Event, { item }) {
  router.push(item.raw.to);
}
</script>

<style scoped>
.dashboard {
  position: relative;
}

/* :deep() selects child components in scoped styles */
.dashboard:deep(tbody tr :hover) {
  cursor: pointer;
}
</style>
