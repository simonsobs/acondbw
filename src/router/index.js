import Vue from "vue";
import VueRouter from "vue-router";
import VueMeta from "vue-meta";

import Frame from "@/components/layout/Frame";
import NullFrame from "@/components/layout/NullFrame";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue"

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import store from "@/store";

Vue.use(VueRouter);
Vue.use(VueMeta);

const routes = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: {
      default: OAuthRedirect,
      frame: NullFrame,
    },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: {
      default: SignIn,
      frame: NullFrame,
    },
  },
  {
    path: "/auth",
    name: "Auth",
    components: {
      default: Auth,
      frame: NullFrame,
    },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: {
      default: SignInError,
      frame: NullFrame,
    },
  },
  {
    path: "/admin/scratch",
    name: "AdminScratch",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/Scratch"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/framework/About.vue"),
  },
  {
    path: "/add-product/:productTypeName",
    components: {
      default: ProductAdd,
      frame: NullFrame,
    },
    meta: { requiresAuth: true }
  },
  {
    path: "/product/:productTypeName",
    components: {
      default: ProductTop,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    children: [
      {
        path: "",
        name: "ProductList",
        component: ProductList,
      },
      {
        path: ":name",
        name: "ProductItem",
        component: ProductItem,
      },
    ],
  },
  {
    path: "*",
    name: "NotFound",
    components: {
      default: () =>
        import(/* webpackChunkName: "error" */ "@/views/framework/NotFound.vue"),
        frame: NullFrame,
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

router.beforeEach((to, from, next) => {
  const authRequired = to.matched.some((record) => record.meta.requiresAuth);
  const signedIn = !!store.state.auth.token;
  if (authRequired) {
    if (signedIn) {
      next();
    } else {
      next("/");
    }
  } else {
    next();
  }
});

export default router;
