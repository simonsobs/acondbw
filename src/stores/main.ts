import { ref, computed } from "vue";
import { defineStore } from "pinia";

export const useStore = defineStore("main", () => {
  const example = ref("abc");
  const snackbar = ref(false);
  const snackbarMessage = ref<string | null>(null);
  const nApolloMutations = ref(0);
  const packageVersion = ref(import.meta.env.PACKAGE_VERSION || "vx.x.x");
  const appVersion = computed(() => packageVersion.value);
  function loadExample() {
    example.value = "123";
  }
  function setSnackbarMessage(message: string) {
    snackbarMessage.value = message;
    snackbar.value = true;
  }
  function closeSnackbar() {
    snackbar.value = false;
  }
  function apolloMutationCalled() {
    nApolloMutations.value++;
  }
  return {
    example,
    snackbar,
    snackbarMessage,
    nApolloMutations,
    packageVersion,
    appVersion,
    loadExample,
    setSnackbarMessage,
    closeSnackbar,
    apolloMutationCalled,
  };
});
