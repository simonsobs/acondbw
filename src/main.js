import Vue from "vue";
import App from "./App.vue";
import router from "./router";
import store from "./store";
import vuetify from "./plugins/vuetify";
import { createProvider, AUTH_TOKEN } from "./vue-apollo";
import { setContext } from "apollo-link-context";

Vue.config.productionTip = false;

// copied from https://blog.logrocket.com/handling-authentication-in-your-graphql-powered-vue-app/
const authLink = setContext(async (_, { headers }) => {
  // get the authentication token from local storage if it exists
  const token = JSON.parse(localStorage.getItem(AUTH_TOKEN));
  // const token = localStorage.getItem(AUTH_TOKEN);
  // Return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      authorization: token || "",
    },
  };
});

console.log(authLink)
const options = { link: authLink };

new Vue({
  router,
  vuetify,
  store,
  apolloProvider: createProvider(options),
  render: (h) => h(App),
  created: function() {
    this.$store.dispatch("setTokenFromLocalStorage");
  }
}).$mount("#app");
