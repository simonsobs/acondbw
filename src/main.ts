import { createApp, h } from "vue";

import pinia from "@/stores";
import { createRouter_ } from "@/router";
import vuetify from "@/plugins/vuetify";
import App from "@/app/App.vue";
import DevToolLoadingStateMenu from "@/components/utils/DevToolLoadingStateMenu.vue";

const router = createRouter_();

const app = createApp({
  render: () => h(App),
});
app.component("DevToolLoadingStateMenu", DevToolLoadingStateMenu);
app.use(pinia);
app.use(router);
app.use(vuetify);
app.mount("#app");
