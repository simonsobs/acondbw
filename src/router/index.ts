import Vue from "vue";
import VueRouter, { RouteConfig } from "vue-router";
import VueMeta from "vue-meta";

import { default as defaultPinia } from "@/stores";
import { useAuthStore } from "@/stores/auth";

import Frame from "@/components/layout/Frame.vue";
import NullFrame from "@/components/layout/NullFrame.vue";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";
import SearchTest from "@/views/framework/SearchTest.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue";

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import ProductType from "@/views/admin/ProductType.vue";

const About = () => import("@/views/framework/About.vue");
const NotFound = () => import("@/views/framework/NotFound.vue");
const AccessDenied = () => import("@/views/framework/AccessDenied.vue");

const FrameAdmin = () =>
  import(/* webpackChunkName: "admin" */ "@/components/admin/FrameAdmin.vue");
const Version = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Version.vue");
const Log = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Log.vue");
const Config = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Config.vue");
const Theme = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Theme.vue");
const User = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/User.vue");
const AdminAppToken = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken.vue"
  );
const AdminAppAuth = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth.vue"
  );
const AdminAppTokenError = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError.vue"
  );

Vue.use(VueRouter);
Vue.use(VueMeta);

let pinia = defaultPinia;

function setPinia(val) {
  // because not sure how to monkey patch pinia for tests
  pinia = val;
}

function setDefaultPinia() {
  setPinia(defaultPinia);
}

const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    // meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/search-test",
    name: "SearchTest",
    components: { default: SearchTest, frame: Frame },
    meta: { requiresAuth: true },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: { default: OAuthRedirect, frame: NullFrame },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: { default: SignIn, frame: NullFrame },
  },
  {
    path: "/auth",
    name: "Auth",
    components: { default: Auth, frame: NullFrame },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: { default: SignInError, frame: NullFrame },
  },
  {
    path: "/admin",
    name: "Admin",
    redirect: { name: "AdminVersion" },
  },
  {
    path: "/admin/product-type",
    name: "ProductType",
    components: { default: ProductType, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/versions",
    name: "AdminVersion",
    components: { default: Version, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/log",
    name: "AdminLog",
    components: { default: Log, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/config",
    name: "AdminConfig",
    components: { default: Config, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/theme",
    name: "AdminTheme",
    components: { default: Theme, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/users",
    name: "AdminUser",
    components: { default: User, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: { default: AdminAppToken, frame: FrameAdmin },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: { default: AdminAppAuth, frame: FrameAdmin },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: { default: AdminAppTokenError, frame: FrameAdmin },
  },
  {
    path: "/about",
    name: "About",
    component: About,
  },
  {
    path: "/add-product/:productTypeName",
    name: "ProductAdd",
    components: { default: ProductAdd, frame: NullFrame },
    meta: { requiresAuth: true },
  },
  {
    path: "/product/:productTypeName",
    components: { default: ProductTop, frame: Frame },
    meta: { requiresAuth: true },
    children: [
      { path: "", name: "ProductList", component: ProductList },
      { path: ":name", name: "ProductItem", component: ProductItem },
    ],
  },
  {
    path: "/access-denied",
    name: "AccessDenied",
    components: { default: AccessDenied, frame: NullFrame },
  },
  {
    path: "*",
    name: "NotFound",
    components: { default: NotFound, frame: NullFrame },
  },
];

function createRouter() {
  const router = new VueRouter({
    mode: "history",
    base: process.env.BASE_URL,
    routes,
  });

  router.beforeEach((to, from, next) => {
    const auth = useAuthStore(pinia);

    const adminRequired = to.matched.some(
      (record) => record.meta.requiresAdmin
    );
    const isAdmin = auth.isAdmin;
    if (adminRequired && !isAdmin) {
      next({ name: "AccessDenied" });
      return;
    }

    const authRequired = to.matched.some((record) => record.meta.requiresAuth);
    const isSignedIn = auth.isSignedIn as boolean;
    if (authRequired && !isSignedIn) {
      next({ name: "AccessDenied" });
      return;
    }
    next();
  });

  return router;
}

const router = createRouter();

function checkAuthForCurrentRoute() {
  const authRequired = router.currentRoute.matched.some(
    (record) => record.meta.requiresAuth
  );
  const auth = useAuthStore(pinia);
  const signedIn = auth.isSignedIn as boolean;
  if (authRequired && !signedIn) {
    router.push("/");
  }
}

export { createRouter, checkAuthForCurrentRoute, setPinia, setDefaultPinia };
