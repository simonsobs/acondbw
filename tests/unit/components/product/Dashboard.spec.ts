import Vue, { ref, nextTick } from "vue";
import VueRouter from "vue-router";
import { PiniaVuePlugin } from "pinia";
import Vuetify from "vuetify";
import { mount, createLocalVue } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";

import Dashboard from "@/components/product/Dashboard.vue";
import { createRouter } from "@/router";

import { useStore } from "@/stores/main";

import { AllProductTypesQuery, ProductTypeEdge } from "@/generated/graphql";

import { useQuery } from "@urql/vue";
jest.mock("@urql/vue");

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("App.vue", () => {
  let localVue: ReturnType<typeof createLocalVue>;
  let vuetify: Vuetify;
  let router: ReturnType<typeof createRouter>;
  let wrapper: ReturnType<typeof mount>;
  let store: ReturnType<typeof useStore>;
  let query: ReturnType<typeof useQuery<AllProductTypesQuery>>;

  const edges: ProductTypeEdge[] = [
    {
      node: {
        id: "UHJvZHVjdFR5cGU6Mw==",
        typeId: "3",
        name: "simulation",
        order: 1,
        indefArticle: "a",
        singular: "simulation",
        plural: "simulations",
        icon: "mdi-creation",
        // @ts-ignore
        products: {
          totalCount: 1,
        },
      },
    },
    {
      node: {
        id: "UHJvZHVjdFR5cGU6MQ==",
        typeId: "1",
        name: "map",
        order: 2,
        indefArticle: "a",
        singular: "map",
        plural: "maps",
        icon: "mdi-map",
        // @ts-ignore
        products: {
          totalCount: 64,
        },
      },
    },
    {
      node: {
        id: "UHJvZHVjdFR5cGU6Mg==",
        typeId: "2",
        name: "beam",
        order: 3,
        indefArticle: "a",
        singular: "beam",
        plural: "beams",
        icon: "mdi-spotlight-beam",
        // @ts-ignore
        products: {
          totalCount: 7,
        },
      },
    },
  ];

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(PiniaVuePlugin);
    vuetify = new Vuetify();
    router = createRouter();
    // @ts-ignore
    query = {
      data: ref<AllProductTypesQuery | undefined>(undefined),
      error: ref(undefined),
      fetching: ref(false),
    };
    (useQuery as jest.Mock).mockReturnValue(query);
    const pinia = createTestingPinia();
    store = useStore(pinia);
    wrapper = mount(Dashboard, {
      localVue,
      vuetify,
      pinia,
      router,
      stubs: ["dev-tool-loading-state-menu"],
    });
    query.data.value = { allProductTypes: { edges: edges } };
  });

  afterEach(() => {
    (useQuery as jest.Mock).mockReset();
  });

  it("match snapshot", async () => {
    await nextTick();
    await nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
