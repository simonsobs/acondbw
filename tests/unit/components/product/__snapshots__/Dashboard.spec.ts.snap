// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue match snapshot 1`] = `
<div class="dashboard" style="position: relative;">
  <div class="v-data-table elevation-1 theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Product type" class="text-start"><span>Product type</span></th>
            <th role="columnheader" scope="col" aria-label="Number of products" class="text-end"><span>Number of products</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-start"><a href="/product/simulation/" class=""><span class="capitalize font-weight-bold primary--text">
          simulations
        </span></a></td>
            <td class="text-end">1</td>
          </tr>
          <tr class="">
            <td class="text-start"><a href="/product/map/" class=""><span class="capitalize font-weight-bold primary--text">
          maps
        </span></a></td>
            <td class="text-end">64</td>
          </tr>
          <tr class="">
            <td class="text-start"><a href="/product/beam/" class=""><span class="capitalize font-weight-bold primary--text">
          beams
        </span></a></td>
            <td class="text-end">7</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <!---->
  <dev-tool-loading-state-menu-stub top="-30px" right="-10px"></dev-tool-loading-state-menu-stub>
</div>
`;
